FAST-LIO/src/laserMapping.cpp

```diff
@@ -162,21 +210,677 @@ ros::Publisher pose_publisher;
+    if (abs(roll) < surroundingkeyframeAddingAngleThreshold &&
+        abs(pitch) < surroundingkeyframeAddingAngleThreshold &&
+        abs(yaw) < surroundingkeyframeAddingAngleThreshold &&
+        sqrt(x * x + y * y + z * z) < surroundingkeyframeAddingDistThreshold)
+        return false;
+    return true;
+}
+
+/**
+ * Add laser odometry factor
+ */
+void addOdomFactor()
+{
+    if (cloudKeyPoses3D->points.empty())
+    {
+        // Initialize a priori factors in the first frame
+        gtsam::noiseModel::Diagonal::shared_ptr priorNoise = gtsam::noiseModel::Diagonal::Variances((gtsam::Vector(6) <<1e-12, 1e-12, 1e-12, 1e-12, 1e-12, 1e-12).finished()); // rad*rad, meter*meter   // indoor 1e-12, 1e-12, 1e-12, 1e-12, 1e-12, 1e-12    //  1e-2, 1e-2, M_PI*M_PI, 1e8, 1e8, 1e8
+        gtSAMgraph.add(gtsam::PriorFactor<gtsam::Pose3>(0, trans2gtsamPose(transformTobeMapped), priorNoise));
+        // Variable node sets initial value
+        initialEstimate.insert(0, trans2gtsamPose(transformTobeMapped));
+    }
+    else
+    {
+        // Add laser odometry factor
+        gtsam::noiseModel::Diagonal::shared_ptr odometryNoise = gtsam::noiseModel::Diagonal::Variances((gtsam::Vector(6) << 1e-6, 1e-6, 1e-6, 1e-4, 1e-4, 1e-4).finished());
+        gtsam::Pose3 poseFrom = pclPointTogtsamPose3(cloudKeyPoses6D->points.back()); /// pre
+        gtsam::Pose3 poseTo = trans2gtsamPose(transformTobeMapped);                   // cur
+        // Parameters: previous frame id, current frame id, pose transformation between the previous frame and the current frame (as an observation value), noise covariance
+        gtSAMgraph.add(gtsam::BetweenFactor<gtsam::Pose3>(cloudKeyPoses3D->size() - 1, cloudKeyPoses3D->size(), poseFrom.between(poseTo), odometryNoise));
+        // Variable node sets initial value
+        initialEstimate.insert(cloudKeyPoses3D->size(), poseTo);
+    }
+}
+
+/**
+ * Add loop closing factor
+ */
+void addLoopFactor()
+{
+    if (loopIndexQueue.empty())
+        return;
+
+    // closed loop queue
+    for (int i = 0; i < (int)loopIndexQueue.size(); ++i)
+    {
+        // The closed loop edge corresponds to the index of the two frames
+        int indexFrom = loopIndexQueue[i].first; //   cur
+        int indexTo = loopIndexQueue[i].second;  //    pre
+        // Pose transformation of closed loop edges
+        gtsam::Pose3 poseBetween = loopPoseQueue[i];
+        gtsam::noiseModel::Diagonal::shared_ptr noiseBetween = loopNoiseQueue[i];
+        gtSAMgraph.add(gtsam::BetweenFactor<gtsam::Pose3>(indexFrom, indexTo, poseBetween, noiseBetween));
+    }
+
+    loopIndexQueue.clear();
+    loopPoseQueue.clear();
+    loopNoiseQueue.clear();
+    aLoopIsClosed = true;
+}
+
+/**
+ * Add GPS factor
+*/
+void addGPSFactor()
+{
+    if (gnss_buffer.empty())
+        return;
+    // If there are no keyframes, or the distance between the first and last keyframes is less than 5m, no GPS factor is added.
```

Comment:
It seems like this only captures the first condition. Should it also be returning if the distance between first/last keyframes is less than 5m, per your comment?

Reply:
Ahh I misread, can you put these 2 conditionals in 1?

Reply:
Aah yes yes